### ISSUE TITLE
Improved or customisable fuzzy match heuristics

### ISSUE THREAD

**Issue by philj56** [philj56/tofi#49]
The [original algorithm](https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/) fuzzy matching is based on has a scoring system designed for searching through code, which isn't tofi's main use case: https://github.com/philj56/tofi/blob/340692299ed834f962b83b1bc0eb6f49471ca556/src/fuzzy_match.c#L170-L176

These scores can occasionally lead to very unintuitive results:
```sh
printf "think\nTxHxIxNxK" | tofi --fuzzy-match true
# Type "think", hit enter
TxHxIxNxK
```

A better default may be to increase the adjacency bonus somewhat, or otherwise tweak the scores.

**Comment by remexre** [comment]
Any reason not to use levenshtein sorting, as rofi does?

**Comment by philj56** [comment]
Not particularly, but I felt that this version of fuzzy matching was a better fit for program launching. Personally, I'm unlikely to make a typo, and very likely to start typing the beginning of a word, which this gives better results for. 

As a completely anecdotal example, here's the results of searching (without history sorting) for "th" in `tofi-run` (left) & `rofi -show run` (right):

![tofi_vs_rofi_th](https://user-images.githubusercontent.com/13610623/213915027-733c4723-3b87-4b6d-ac4e-44910585f1a8.png)

Out of rofi's first page of results, I'd only consider "Thunar" to be useful - there's no chance I was looking for any of those other results. Tofi's results seem much more sensible to me, with the first page full of words starting in "th".

If we then extend our search to "thin", we get:

![tofi_vs_rofi_thin](https://user-images.githubusercontent.com/13610623/213915349-d02822ad-d9c6-478b-9c45-61ae4572c668.png)

Even now, rofi prefers "cowthink" as a match over all of the commands starting with "thin".

As people are using this for more than just program launching though, it might be nice to add an option to use Levenshtein distance (or some other more general, typo-resistant metric). I'll try to have a look into it at some point.

**Comment by philj56** [comment]
Not particularly, but I felt that this version of fuzzy matching was a better fit for program launching. Personally, I'm unlikely to make a typo, and very likely to start typing the beginning of a word, which is why I don't use the fuzzy matching much myself. 

As a completely anecdotal example, here's the results of searching (without history sorting) for "th" in `tofi-run` (left) & `rofi -show run` (right):

![tofi_vs_rofi_th](https://user-images.githubusercontent.com/13610623/213916093-8495e895-9e35-460d-b5b2-cf0cc471a250.png)

Out of rofi's first page of results, I'd only consider "Thunar" to be useful - there's no chance I was looking for any of those other results. Tofi's results seem much more sensible to me, with the first page full of words starting in "th", and the top result being an initialism (though see my original comment).

If we then extend our search to "thin", we get:

![tofi_vs_rofi_thin](https://user-images.githubusercontent.com/13610623/213916106-57ff72a4-a396-40a4-8d78-aadc6085a4a3.png)

Tofi has pretty much reverted to the no fuzzy matching case, giving results starting with "thin". Rofi's results seem pretty much useless to me, with only "cowthink" and "thin_check" being viable - it's far too forgiving of differences. 

As people are using this for more than just program launching though, it might be nice to add an option to use Levenshtein distance (or some other more general, typo-resistant metric). I'll try to have a look into it at some point.

**Comment by remexre** [comment]
Ah, that makes sense; and yeah, for program launching, I've had no complaints.

The usecase where I've wanted Levenshtein distance is as a frontend for the [pass](https://passwordstore.org) password manager; I store my passwords hierarchically (e.g. `Personal/Dev/GitHub/remexre`), so usually I'm starting typing at the second-to-last component.

**Comment by philj56** [comment]
Yeah, that seems to be a common use case, although I have to admit I haven't tried it myself despite using pass.

As a side note, your hierarchy actually seems like it should lend itself well to the current fuzzy matching, as initialisms are strongly rewarded. For `Personal/Dev/GitHub/remexre`, typing `pdgr`  gives an extremely high score - in fact it's higher than matching `pdgr` itself! The downside I guess is that it's more awkward to have to think about typing like this.

**Comment by philj56** [comment]
I spent a while implementing Levenshtein distance (actually [OSA](https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance#Optimal_string_alignment_distance)) as a matching algorithm, but as far as I can tell it doesn't really suit a menu program like tofi. The problem is that it's only a good metric when you've typed almost the whole target string; it's not very good at matching substrings.

Maybe it'd work integrated with some other matching logic, e.g. "match substrings with a Levenshtein distance of 2 or less", but that's more complex.

As an aside, rofi seems to do something weird with its fuzzy matching, seemingly filtering results based on "are all these letters here in the correct order", then sorting those results based on Levenshtein distance. However, this means no typos (substitutions or transpositions) are allowed, so the "Levenshtein" distance is actually just the difference in string length! Or at least it should be, but that's not what the output looks like - I don't know what's actually going on there.

**Comment by herrvonvoid** [comment]
hi, this seems to be pretty close to the issue i have. could we have a `--fuzzy-prefer-exact` switch or something similar, because i have the following "problem": if i type reboot the results are
1. repack_bootimg
2. reboot
3. grub-reboot

and i would prefer to have "reboot" as the first result
